### DISCUSSION TITLE
Very low map in segementation training

### DISCUSSION THREAD

**Discussion by someuser2026** [ultralytics/ultralytics#21299] [Q&A] [answered]
I am training custom models with backbones from torchvision on a custom dataset, on instance segmentation. I three models based on convnext base, without any neck, same neck as yolov11 and a custom head similar to yolov11 but additional C3k2 block to combine even higher res feature maps. The first two work fine with ok map for masks however the third one has very low mAP. Attached are the logs during training as well as the yamls for all the models. What could be the reason for it?

Note: The other two models also lead to nans in losses after 250+ epochs

Thank you. Greatly appreciate your time.

```

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
     16/478      18.2G      1.343      2.019      1.443      2.228         26        384: 100%|██████████| 62/62 [00:07<00:00,  8.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  9.10it/s]
                   all        109        264      0.797      0.683      0.768      0.335    0.00855    0.00379   0.000323   9.21e-05

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
     17/482      18.2G      1.331      2.015      1.396      2.145         30        128: 100%|██████████| 62/62 [00:07<00:00,  8.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  9.08it/s]
                   all        109        264       0.85      0.682      0.809      0.373    0.00471    0.00379   0.000195   5.66e-05

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
     18/481      18.2G       1.33      2.016      1.435      2.216         19        288: 100%|██████████| 62/62 [00:07<00:00,  8.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  9.14it/s]
                   all        109        264      0.879      0.705      0.816      0.386    0.00348    0.00758   0.000255   6.52e-05

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
     19/479      18.2G      1.326      1.991      1.448      2.196         19        224: 100%|██████████| 62/62 [00:07<00:00,  7.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  9.10it/s]
                   all        109        264      0.827      0.724      0.808      0.349    0.00511    0.00379   0.000248   9.41e-05

```

### No neck
```
nc: 1
backbone:
  - [-1, 1, TorchVision, [1024, "convnext_base", "DEFAULT", True, 2, True]]  # - 0
head:
  - [0, 1, Index, [128, 2]] # - 1
  - [0, 1, Index, [256, 4]] # - 2
  - [0, 1, Index, [512, 6]] # - 3
  - [0, 1, Index, [1024, 8]] # - 4
  - [[2, 3, 4], 1, Segment, [nc, 32, 256]] # pass all three feature to obb head
```

### YOLOv11 neck
```
# convnext base, head and neck same as yolo11
# 0 torch.Size([1, 3, 640, 640])
# 1 torch.Size([1, 128, 160, 160])
# 2 torch.Size([1, 128, 160, 160])
# 3 torch.Size([1, 256, 80, 80])
# 4 torch.Size([1, 256, 80, 80])
# 5 torch.Size([1, 512, 40, 40])
# 6 torch.Size([1, 512, 40, 40])
# 7 torch.Size([1, 1024, 20, 20])
# 8 torch.Size([1, 1024, 20, 20])

nc: 1
scales:
  x: [1.00, 1.00, 1024]
backbone:
  - [-1, 1, TorchVision, [1024, "convnext_base", "DEFAULT", True, 2, True]]  # - 0
head:
  - [0, 1, Index, [128, 2]] # - 1
  - [0, 1, Index, [256, 4]] # - 2
  - [0, 1, Index, [512, 6]] # - 3
  - [0, 1, Index, [1024, 8]] # - 4

  - [4, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 3], 1, Concat, [1]] # cat backbone P4
  - [-1, 2, C3k2, [512, False]] # 7

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 2], 1, Concat, [1]] # cat backbone P3
  - [-1, 2, C3k2, [256, False]] # 10 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 7], 1, Concat, [1]] # cat head P4
  - [-1, 2, C3k2, [512, False]] # 13 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 4], 1, Concat, [1]] # cat head P5
  - [-1, 2, C3k2, [1024, True]] # 16 (P5/32-large)

  - [[10, 13, 16], 1, Segment, [nc, 32, 256]] # pass all three feature to obb head
```

### Customized YOLOv11 neck
```
# convnext base, head and neck same as yolo11
# 0 torch.Size([1, 3, 640, 640])
# 1 torch.Size([1, 128, 160, 160])
# 2 torch.Size([1, 128, 160, 160])
# 3 torch.Size([1, 256, 80, 80])
# 4 torch.Size([1, 256, 80, 80])
# 5 torch.Size([1, 512, 40, 40])
# 6 torch.Size([1, 512, 40, 40])
# 7 torch.Size([1, 1024, 20, 20])
# 8 torch.Size([1, 1024, 20, 20])

nc: 1
scales:
  x: [1.00, 1.00, 1024]
backbone:
  - [-1, 1, TorchVision, [1024, "convnext_base", "DEFAULT", True, 2, True]]  # - 0
head:
  - [0, 1, Index, [128, 2]] # - 1
  - [0, 1, Index, [256, 4]] # - 2
  - [0, 1, Index, [512, 6]] # - 3
  - [0, 1, Index, [1024, 8]] # - 4

  - [4, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 3], 1, Concat, [1]] # cat backbone P4
  - [-1, 2, C3k2, [512, False]] # 7

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 2], 1, Concat, [1]] # cat backbone P3
  - [-1, 2, C3k2, [256, False]] # 10 (P3/8-small)

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 1], 1, Concat, [1]] # cat backbone P3
  - [-1, 2, C3k2, [256, False]] # 13 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 10], 1, Concat, [1]] # cat head P4
  - [-1, 2, C3k2, [512, False]] # 16 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 7], 1, Concat, [1]] # cat head P5
  - [-1, 2, C3k2, [1024, True]] # 19 (P5/32-large)

  - [[13, 16, 19], 1, Segment, [nc, 32, 256]] # pass all three feature to obb head
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @someuser2026, thank you for sharing your detailed setup and results with Ultralytics 🚀! We appreciate your effort in providing model YAMLs and training logs. An Ultralytics engineer will review your discussion and assist you soon.

If this is a 🐛 Bug Report, please also provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

For custom training ❓ questions, the more context you provide (such as representative dataset images, additional config details, and any relevant logs), the easier it is for us to help. Make sure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for optimal performance.

Join the Ultralytics community where it suits you best:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

😊 Thank you for your patience and for helping to improve Ultralytics!

**Reply by glenn-jocher** [reply, marked as answer]
The issue with your custom neck is that you're using P2 features (layer 13) which are too high-resolution and noisy for effective segmentation. Your custom model uses features from layers [13, 16, 19] representing P2/4, P3/8, P4/16 scales, while the working YOLOv11 neck uses [10, 13, 16] representing P3/8, P4/16, P5/32 scales.

P2 features at 4x downsampling are generally too detailed and contain excessive noise for robust segmentation learning, which explains the near-zero mask mAP values. Try modifying your custom neck to use the standard P3, P4, P5 feature scales like the working YOLOv11 version, or experiment with P2, P3, P4 if you need higher resolution segmentation, but ensure proper feature processing and channel alignment throughout the neck.

**Reply by someuser2026** [reply]
Thank you! Appreciate the help. I have some very small and ambiguous objects in the images with edges blending in with the surroundings. I thought including high res feature might work.

**Reply by glenn-jocher** [reply]
That's understandable reasoning for small objects! For tiny and ambiguous objects, try these approaches instead: increase your input image size during training (e.g., `imgsz=1024` or higher), use more aggressive data augmentation like `mosaic=1.0` and `mixup=0.15`, or consider training with smaller anchor scales. You could also experiment with the standard P3/P4/P5 setup but add more C3k2 layers or increase the depth multiplier to give the model more capacity to learn fine details without the noise issues of P2 features.

**Reply by someuser2026** [reply]
Hi! I am already using imgs=1024 and mosiac=1. However, beacuse of  instability during training leading to nan loss after around 100 epochs, i used close_mosiac=10 as mentioned in [this discussion](https://github.com/orgs/ultralytics/discussions/20083#discussioncomment-13686916)

**Reply by glenn-jocher** [reply]
Good approach with `close_mosaic=10` for training stability! For small ambiguous objects with stable training, try increasing the model's capacity by using a larger variant (e.g., `yolo11m-seg.pt` instead of `yolo11s-seg.pt`) or adding more layers to your neck architecture, and consider using `copy_paste=0.3` augmentation which is particularly effective for segmentation tasks with small objects.

**Reply by someuser2026** [reply]
Thank you! I am using yolov11x as baseline. Would look inot the copy_paste. Also planning to see if cropped inference yeilds better results. Appreciate the help!

**Reply by glenn-jocher** [reply]
You're welcome! YOLOv11x is an excellent choice for challenging small object segmentation. Cropped inference with sliding windows or test-time augmentation can indeed improve results significantly for tiny objects - you might also consider using `predict()` with `augment=True` for built-in TTA during inference. Good luck with your project!
